### PR TITLE
Fix immutable color cache

### LINKED ISSUE #2329: "Attempting to assign to readonly property" error when using SharedValue across Runtimes (VisionCamera) (interpolateColor: `getInterpolateCacheRGBA`)
## Description

I'm using this hook:

```tsx
export function useAnimatedColor(
  color: Reanimated.SharedValue<string>,
): Readonly<Reanimated.SharedValue<string | number>> {
  const animation = useSharedValue(0);
  const colorFrom = useSharedValue(DEFAULT_COLOR);
  const colorTo = useSharedValue(color.value);

  useAnimatedReaction(
    () => color.value,
    (newColor, prevColor) => {
      animation.value = 0;
      colorFrom.value = prevColor ?? DEFAULT_COLOR;
      colorTo.value = newColor;
      animation.value = withTiming(1, {
        duration: 150,
        easing: Easing.linear,
      });
      console.log(`Animating from ${prevColor} -> ${newColor}`);
    },
  );

  // TODO: Using colorFrom and colorTo in here raises "Attempting to assign to readonly property" error...
  return useDerivedValue(() =>
    interpolateColor(animation.value, [0, 1], [colorFrom.value, colorTo.value]),
  );
}
```

which results in this error being thrown:

### Expected behavior

I expect it to interpolate correctly

### Actual behavior & steps to reproduce

It crashes

## Snack or minimal code example

https://github.com/mrousavy/Colorwaver ([this code](https://github.com/mrousavy/Colorwaver/blob/40bccf92f1533f08be7be4e425f3fa3b8b30ed28/src/useAnimatedColor.ts#L32-L35) throws the error. the passed in SharedValue is being assigned [here](https://github.com/mrousavy/Colorwaver/blob/40bccf92f1533f08be7be4e425f3fa3b8b30ed28/src/App.tsx#L37-L44))

## Package versions

- React Native: 0.65.1
- React Native Reanimated: 2.3.0-alpha.2
- NodeJS: 14
- Xcode: /
- Java & Gradle: /

## Affected platforms

- [ ] Android (not tested)
- [x] iOS
- [ ] Web (not tested)

### PR DESCRIPTION
**Pull request by piaskowyk**
## Description

The color cache was mutable only from the JS side. I decided to remove the cache and instead of this I created a new faster function with a better cache mechanism without previous limitations.

Performance comparation between `interpolateColor` and `interpolateSharableColor`:
```
old 2894ms - new 1741ms // 39.8% faster
old 2720ms - new 1697ms // 37.6% faster
old 2739ms - new 1650ms // 39.7% faster
old 2738ms - new 1781ms // 34.9% faster
old 2727ms - new 1662ms // 39.0% faster
```
Performance test code:
```js
const interpolateConfig = useInterpolateConfig([0, 350], ['#ff0000', '#00ff00']);
const v1 = interpolateSharableColor(randomWidth.value, interpolateConfig);
const iterations = 10000;
const t1_start = Date.now();
for(let i = 0; i < iterations; i++) {
  const v = interpolateSharableColor(randomWidth.value, interpolateConfig);
}
const t1_end = Date.now();
const t2_start = Date.now();
for(let i = 0; i < iterations; i++) {
  const v = interpolateColor(randomWidth.value, [0, 350], ['#ff0000', '#00ff00']);
}
const t2_end = Date.now();
console.log(t1_end - t1_start, t2_end - t2_start);
```

## Usage
```js
import Animated, {
  useSharedValue,
  withTiming,
  useAnimatedStyle,
  interpolateSharableColor,
  useInterpolateConfig,
} from 'react-native-reanimated';
import { View, Button } from 'react-native';
import React from 'react';

function AnimatedStyleUpdateExample(): React.ReactElement {
  const randomWidth = useSharedValue(10);
  const interpolateConfig = useInterpolateConfig([0, 350], ['#ff0000', '#00ff00']); // <- here

  const style = useAnimatedStyle(() => {
    return {
      width: randomWidth.value,
      backgroundColor: interpolateSharableColor(randomWidth.value, interpolateConfig) // <- here
    };
  });

  return (
    <View style={{ flex: 1, flexDirection: 'column', }}>
      <Animated.View
        style={[ 
          { width: 100, height: 80, backgroundColor: 'black', margin: 30 }, 
          style 
        ]}
      />
      <Button
        title="toggle"
        onPress={() => { randomWidth.value = withTiming(Math.random() * 350) }}
      />
    </View>
  );
}

export default AnimatedStyleUpdateExample;
```

Fixes #2329 #2739

### CODE REVIEW

File: src/reanimated2/Colors.ts [hunk 1]
```diff
@@ -759,13 +767,17 @@ const getInterpolateCacheHSV = (
...
-  interpolateCacheHSV[hash] = newCache;
-  hashOrderHSV[curentHashIndexHSV] = hash;
-  curentHashIndexHSV = (curentHashIndexHSV + 1) % BUFFER_SIZE;
+  const mutableCache = {
+    ...interpolateCacheHSV.value,
+  };
+  mutableCache[hash] = newCache;
+  interpolateCacheHSV.value = mutableCache;
+  hashOrderHSV.value[curentHashIndexHSV.value] = hash;
+  curentHashIndexHSV.value = (curentHashIndexHSV.value + 1) % BUFFER_SIZE;
```
Comment (tomekzaw) [inline src/reanimated2/Colors.ts:780]: With `BUFFER_SIZE = 1` this is always 0.

File: src/reanimated2/Colors.ts [hunk 2]
```diff
@@ -685,17 +685,19 @@ interface InterpolateCacheRGBA {
   a: number[];
 }
 
-const BUFFER_SIZE = 200;
-const hashOrderRGBA: any = new ArrayBuffer(BUFFER_SIZE);
-let curentHashIndexRGBA = 0;
-const interpolateCacheRGBA: { [name: string]: InterpolateCacheRGBA } = {};
+const BUFFER_SIZE = 1;
```
Comment (tomekzaw) [inline src/reanimated2/Colors.ts:688]: Is this change necessary?

File: src/reanimated2/Colors.ts [hunk 3]
```diff
@@ -716,13 +718,17 @@ const getInterpolateCacheRGBA = (
...
-  const overrideHash = hashOrderRGBA[curentHashIndexRGBA];
+  const overrideHash = hashOrderRGBA.value[curentHashIndexRGBA.value];
   if (overrideHash) {
     delete interpolateCacheRGBA[overrideHash];
   }
-  interpolateCacheRGBA[hash] = newCache;
-  hashOrderRGBA[curentHashIndexRGBA] = hash;
-  curentHashIndexRGBA = (curentHashIndexRGBA + 1) % BUFFER_SIZE;
+  const mutableCache = {
+    ...interpolateCacheRGBA.value,
```
Comment (tomekzaw) [inline src/reanimated2/Colors.ts:726]: It looks like here we're unpacking the whole object (effectively cloning it) but previously we were just assigning a single value to a single key.

Comment (piaskowyk) [reply src/reanimated2/Colors.ts:726]: I agree but this object is unfortunately frozen and this is the only way to modify it.

**Comment by piaskowyk** [comment]
Sorry, this approach is wrong, but I have a much better idea!

**Comment by piaskowyk** [comment]
@tomekzaw What do you think about this?

File: src/reanimated2/Colors.ts [hunk 4]
```diff
@@ -780,16 +746,73 @@ export const interpolateColor = (
...
+      interpolateConfig.value.cache.value = colors;
+    }
+    return interpolateColorsHSV(
+      value,
+      interpolateConfig.value.inputRange,
+      colors as InterpolateHSV
+    );
+  }
+  throw new Error(
+    `invalid color space provided: ${interpolateConfig.value.colorSpace}. Supported values are: ['RGB', 'HSV']`
```
Comment (tomekzaw) [inline src/reanimated2/Colors.ts:816]:
```suggestion
    `Invalid color space provided: ${interpolateConfig.value.colorSpace}. Supported values are: ['RGB', 'HSV']`
```

File: src/reanimated2/Colors.ts [hunk 5]
```diff
@@ -780,16 +746,73 @@ export const interpolateColor = (
     return interpolateColorsHSV(
       value,
       inputRange,
-      getInterpolateCacheHSV(outputRange)
+      getInterpolateHSV(outputRange)
     );
   } else if (colorSpace === 'RGB') {
```
Comment (tomekzaw) [inline src/reanimated2/Colors.ts:751]: Shouldn't it be RGBA instead?

**Review by tomekzaw — APPROVED** [review]

**Pull request merged**